✨EVM

raxhvl opened the pull request:
Placeholder PR to communicate that I'll be working on EVM. Hoping to finish by the end of this week.

raxhvl commented:
Hey @taxmeifyoucan I have completed the write up on EVM. [🔗 Preview link](https://epf-wiki-git-feat-evm-0xrahl.vercel.app/#/wiki/EL/evm)

@lightclient: It would be awesome if you could review it too.

taxmeifyoucan commented on docs/wiki/EL/evm.md:
```diff
@@ -0,0 +1,623 @@
+
+In computer science, **state machine** is an abstraction used to model the behavior of a system. It illustrates how a system can be represented by a set of distinct states and how inputs can drive changes in the state.
+
+A familiar is a vending machine, an automated system for dispensing products upon receiving payment.
```
I guess you meant 'A familiar examples.."?

raxhvl replied:
Yep 😅

taxmeifyoucan commented on docs/wiki/EL/evm.md:
```diff
@@ -0,0 +1,623 @@
+
+![EVM Gas](../../images/evm/evm-gas.gif)
+
+Our example assumed 1 unit of gas per opcode for simplicity, but the actual cost varies based on complexity. The core concept, however, remains unchanged.
```
you can elaborate a tiny bit here, that the gas cost are measured to reflect the constrains of hardware - storage or computation

taxmeifyoucan commented on docs/wiki/EL/evm.md:
```diff
@@ -0,0 +1,623 @@
+
+- **nonce ($T_n$)**: An integer value equal to the number of transactions sent by the sender. Nonce is used to:
+
+  - **Prevent replay attack**: Let's say Alice sends 1 ETH to Bob in a transaction, Bob might try to rebroadcast the same transaction into the network to get additional funds from Alice's account. Since the transaction is signed with a unique nonce, EVM will simply reject it if Bob sends it again. Thus safeguarding Alice's account from unauthorized duplicate transactions.
```
There is a doc on transaction anatomy worked on #67. It needs a *a LOT* of polish right now but I would suggest moving some of this content there because only certain fields are relevant here

raxhvl replied:
I would prefer if we moved the whole transaction section to #67 and then link it here. I had to include it here because some context is required but its actually a tangential topic. Does that sound okay?

taxmeifyoucan replied:
Sounds perfect!

taxmeifyoucan commented on docs/wiki/EL/evm.md:
I would mention that this is an example using foundry as one of the tools which can be used for smart contract development and deployment

taxmeifyoucan reviewed (commented):
Great content, thanks a lot @raxhvl! I left few comments about minor issues, otherwise it's ready for merging

lightclient commented on docs/wiki/EL/evm.md:
```diff
@@ -0,0 +1,623 @@
+# The Ethereum Virtual Machine (EVM)
+
+You have likely observed that Ethereum transactions take some time to finalise, [around 12 seconds](https://web.archive.org/web/20240304145956/https://etherscan.io/chart/blocktime) at the time of writing. A series of crucial steps occur during this timeframe. The transaction is first queued in a pool. Then, it is selected by a node and executed by a special program - **the Ethereum Virtual Machin ...
```
also I would note that the block is not finalized after 12s -- 12s is just the slot time for the beacon chain. there may or may not be a block in a given slot. finalization happens after 2 epochs (~12 min)

raxhvl replied:
Jeez i need to step up the intro game!

How does this simple intro sound:

The Ethereum Virtual Machine (EVM) is the heart of the Ethereum World Computer. It performs the crucial computations needed to finalise transactions, permanently storing the results on the blockchain. This article explores the EVM's role in the Ethereum ecosystem and how it works.

raxhvl commented:
@taxmeifyoucan Thanks for the review. I have made the following updates:

- [x] Simplified introduction.
- [x] Relevant placing for explanation of global state and link to data structures.
- [x] Examples of popular VMs.
- [x] Brief about how clients implement standard EVM specs.
- [x] Primer about gas cost calculation.
- [x] Brief introduction to foundry.
- [x] Move transaction to #67. 
- [x] Fixed typos, and link to JSON-RPC.

taxmeifyoucan commented:
Looks great, thanks a lot @raxhvl!
